### PR TITLE
Add a global exception handler to call MPI_Abort

### PR DESCRIPTION
**Pull request by keisukefukuda**
This PR implements a trick to hook a global exception handler to call `MPI_Abort()`.
The motivation is discussed in #236 .

A new environmental variable `CHAINERMN_FORCE_ABORT_ON_EXCEPTION` is introduced.
The design and implementation are still subject to discuss.

This PR is tested to work properly using `train_mnist.py` .
The issue is reproduced by changing `train_mnist.py` as:

```python
    # Some display and output extensions are necessary only for one worker.
    # (Otherwise, there would just be repeated outputs.)
    if comm.rank == 0:
        raise ValueError('error')    # <---- Inserted
        trainer.extend(extensions.dump_graph('main/loss'))
        trainer.extend(extensions.LogReport())
        trainer.extend(extensions.PrintReport(
            ['epoch', 'main/loss', 'validation/main/loss',
             'main/accuracy', 'validation/main/accuracy', 'elapsed_time']))
        trainer.extend(extensions.ProgressBar())
```

### CODE REVIEW

File: chainermn/__init__.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from chainermn import iterators  # NOQA
 from chainermn import links  # NOQA
 from chainermn import optimizers  # NOQA
+import chainermn.util  # NOQA
```
Comment (kuenishi) [inline chainermn/__init__.py]: This case so specific that the file name should be specific like global-hook or global-exception-hook, or something like that.

File: docs/source/tutorial/tips_faqs.rst [hunk 2]
```diff
@@ -39,3 +39,38 @@ Please refer to `MPI4py API reference <http://pythonhosted.org/mpi4py/apiref/mpi
 Using FP16
 ~~~~~~~~~~
 FP16 (16-bit half precision floating point values) is not supported in ChainerMN as of now.
```
Comment (kuenishi) [inline docs/source/tutorial/tips_faqs.rst]: This faq is duplicate with #249 and will conflict with it.

File: chainermn/util.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        except Exception as e:
+            # Something is completely broken...
+            # There's nothing we can do any more ¯\_(ツ)_/¯
+            sys.stderr.write(
+                "Sorry, failed to stop MPI and the process may hang.\n")
+            sys.stderr.flush()
+            raise e
+
+
+#
```
Comment (kuenishi) [inline chainermn/util.py]: Code comments below look duplicate with additional faq? More specific comment would be nice as docstring for this function added to docs in `docs/source/reference/index.rst` .

File: chainermn/__init__.py (hunk 1 above)
Comment (keisukefukuda) [reply chainermn/__init__.py]: I rename the file to `global_exc_hook.py`.

File: docs/source/tutorial/tips_faqs.rst (hunk 2 above)
Comment (keisukefukuda) [reply docs/source/tutorial/tips_faqs.rst]: Fixed!

File: chainermn/util.py (hunk 3 above)
Comment (keisukefukuda) [reply chainermn/util.py]: I added a `configuration` section to the reference manual, although there are only 1 entry of `CHAINERMN_FORCE_ABORT_ON_EXCEPTION` .

File: chainermn/global_exc_hook.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+# It is problematic especially when you run ChainerMN programs on a cloud
+# environment, on which you are charged on time basis.
+# See https://github.com/chainer/chainermn/issues/236 for more discussion.
+#
+# To activate this handler, set CHAINERMN_FORCE_ABORT_ON_EXCEPTION
+# to a non-empty value.
+# Note that you need to pass an argument to mpiexec (-x for Open MPI)
+# to activate the handler in all processes.
+#
+def add_hook_if_enabled():
```
Comment (kuenishi) [inline chainermn/global_exc_hook.py]: If this should not be exposed as official API of ChainerMN, the name should start with `_` like `_add_hook_if_enabled():` .

File: chainermn/global_exc_hook.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    var = os.environ.get('CHAINERMN_FORCE_ABORT_ON_EXCEPTION')
+    if var is not None and len(var) > 0:
+        add_hook()
+
+
+def add_hook():
```
Comment (kuenishi) [inline chainermn/global_exc_hook.py]: I think this is an officially exposed function, so `autofunction::` should be enabled at API docs.

Comment (kuenishi) [reply chainermn/global_exc_hook.py]: Also it should have reasonable docstring.

File: chainermn/global_exc_hook.py (hunk 4 above)
Comment (keisukefukuda) [reply chainermn/global_exc_hook.py]: Fixed. Thanks.

File: chainermn/global_exc_hook.py (hunk 5 above)
Comment (keisukefukuda) [reply chainermn/global_exc_hook.py]: Fixed. Thanks

File: docs/source/reference/index.rst [hunk 6]
```diff
@@ -9,10 +9,9 @@ Communicators
 
 .. autofunction:: create_communicator
 .. autoclass:: CommunicatorBase
-    :members: rank, intra_rank, inter_rank, inter_size, size,
-              alltoall, split, send, recv, bcast, gather, allreduce,
-              send_obj, recv_obj, bcast_obj, gather_obj,
-              allreduce_obj, bcast_data, allreduce_grad
+    :members: rank, intra_rank, size, alltoall, split, send, recv,
+              bcast, gather, allreduce, send_obj, recv_obj, bcast_obj,
+              gather_obj, allreduce_obj, bcast_data, allreduce_grad
```
Comment (kuenishi) [inline docs/source/reference/index.rst]: I don't know why this change happens but this is a rollback of https://github.com/chainer/chainermn/commit/f6c21a9e8d6164bac8c86b36c432530fa7356235 and should not happen.

File: docs/source/reference/index.rst [hunk 7]
```diff
@@ -43,11 +42,8 @@ Functions
 .. autofunction:: chainermn.functions.send
 .. autofunction:: chainermn.functions.recv
 .. autofunction:: chainermn.functions.pseudo_connect
+.. autofunction:: chainermn.functions.all_to_all
 .. autofunction:: chainermn.functions.bcast
-.. autofunction:: chainermn.functions.gather
-.. autofunction:: chainermn.functions.scatter
-.. autofunction:: chainermn.functions.alltoall
-.. autofunction:: chainermn.functions.allgather
```
Comment (kuenishi) [inline docs/source/reference/index.rst]: Ditto, all of these functions should be exposed as ChainerMN's API.

**Comment by keisukefukuda** [comment]
Thanks for the comment.
I checked my commit `0078a6b1`, but I had no idea why the problematic change was made.
Anyways, I fixed the issue.

**Comment by keisukefukuda** [comment]
The `make html` issue is not yet fixed. I'm working on that.

**Comment by keisukefukuda** [comment]
Now I improved the function docstring to point to the FAQ entry, which explains the problem and two possible solutions (using `-m mpi4py` and using `add_hook()`)
